File: src/main/java/org/apache/prepbuddy/smoothers/Weights.java
package org.apache.prepbuddy.smoothers;

import org.apache.prepbuddy.exceptions.ApplicationException;
import org.apache.prepbuddy.exceptions.ErrorMessages;

import java.io.Serializable;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;

/**
 * Contains weights in sequence for the weighted sliding window.
 */
public class Weights implements Serializable {
    private int limit;
    private List<Double> weights;

    public Weights(int limit) {
        this.limit = limit;
        weights = new ArrayList<>(this.limit);
    }

    public void add(double value) {
        if (size() == limit)
            throw new ApplicationException(ErrorMessages.SIZE_LIMIT_IS_EXCEEDED);
        if (size() == limit - 1 && sumWith(value) != 1.0)
            throw new ApplicationException(ErrorMessages.WEIGHTS_SUM_IS_NOT_EQUAL_TO_ONE);
        weights.add(value);
    }

    public double get(int index) {
        return weights.get(index);
    }

    private double sumWith(double value) {
        Double sum = 0.0;
        for (Double oneValue : weights) {
            sum += oneValue;
        }
        sum += value;
        return Double.parseDouble(new DecimalFormat("##.#").format(sum));
    }

    public int size() {
        return weights.size();
    }
}

File: src/main/scala/org/apache/datacommons/prepbuddy/smoothers/Weights.scala
package org.apache.datacommons.prepbuddy.smoothers

import org.apache.datacommons.prepbuddy.exceptions.{ApplicationException, ErrorMessages}

/**
  * Contains weights in sequence for the weighted sliding window.
  */
class Weights(limit: Int) extends Serializable {
    private var weights: List[Double] = List()

    def multiplyWith(list: List[Double]): List[Double] = {
        weights.zip(list).map((tuple) => tuple._1 * tuple._2)
    }

    def get(index: Int): Double = weights(index)

    private def sumWith(value: Double): Double = "%1.1f".format(weights.sum + value).toDouble

    def add(value: Double): Unit = {
        if (size == limit) throw new ApplicationException(ErrorMessages.SIZE_LIMIT_IS_EXCEEDED)
        if(size == limit - 1 && sumWith(value) != 1.0 ){
            throw new ApplicationException(ErrorMessages.WEIGHTS_SUM_IS_NOT_EQUAL_TO_ONE)
        }
        weights = weights :+ value
    }

    def size : Int = weights.length
}
